cloudevents/working-drafts/minor-versions.md

```diff
@@ -0,0 +1,337 @@
+
+An event consumer that doesn't know about v1.1 should still be able to consume
+a v1.1 event without any knowledge of the `xyz` attribute; it should be treated as a regular
+extension attribute. However, the consumer does "know" that the event is v1.1, so can be aware
+that `xyz` (and any other extension attributes) *may* be a core attribute in v1.1.
+(In other words, they can distinguish between `xyz` appearing in a v1.0 event which is *definitely*
+an extension attribute as the consumer is aware of all core v1.0 attribtes,
+and `xyz` appearing in a v1.1 event which is *either* an extension attribute *or* a new optional
+core attribute.)
+
+## SDKs
+
+If we adopt minor versions, SDKs should change to support them, so that (for example) an SDK
+supporting v1.1 will *also* parse v1.2 events. We should keep track of which SDKs
+(and which versions of those SDKs) support parsing v1.x events.
+
+Different SDKs expose features very differently. This document is written from the perspective
+of the C# SDK, but other viewpoints would be highly valued: please add feedback, particularly
+if the approach suggested doesn't make sense for an SDK you're familiar with.
+
+### Creating events: default version handling
+
+It may sound natural to allow something like:
+
+```csharp
+var evt = new CloudEvent { Id = "..." };
+```
+
+... to just use the latest spec version that the SDK supports. However, this is a breaking change.
+Consider:
+
+```csharp
+var evt = new CloudEvent
+{
+    Id = "...",
+    ["xyz"] = "test"
+};
+```
+
+When the SDK only supports v1.0, that creates a valid event. When the SDK supports v1.1 as well,
+that would produce an invalid event or throw an exception, if the version is implicitly v1.1.
+Instead, the SDK should *implicitly* use the latest version that was present when the first stable
+version of the SDK came out, within its own current major version. In other words, if v1.1 ships
+and a new SDK is created, that can default to v1.1 - and if an SDK creates its own new major version,
+*that* can default to v1.1 as well... but in both cases that default should stay as v1.1 in new minor
+SDK versions even after v1.2 ships.
+
+An SDK *may* choose to provide a way of explicitly opting into the "latest supported" version:
+
+```csharp
+var evt = new CloudEvent(CloudEventsSpecVersion.Latest)
+{
+  
+};
+```
+
+### Accessing attributes
+
+If an SDK exposes core attributes in a distinct way from extension attributes (as shown above,
+where `Id` is a strongly-typed property, whereas `["xyz"]` uses an indexer with a string literal)
+then any attempt to use the "special" access should fail when accessing an attribute within
+an event with a version that doesn't have that attribute as a core attribute. Example:
+
+```csharp
+// Create a v1.0 event
+var evt = new CloudEvent(CloudEventsSpecVersion.V1_0);
+
+// Try to access the new Xyz property, referring to the optional attribute introduced in
+// v1.1: this should fail.
+evt.Xyz = 1;
```

Comment:
I actually disagree that it should fail. If the SDK has explicitly decided to support an extension as a first class thing then I think this should work.

Reply:
> But it means something in 1.1 and not in 1.0

This is a v1.0 CE so the SDK is safe, and allowed, to define their own attributes. The fact it MIGHT mean something in a future CE spec shouldn't impact what it does for v1.0.

Now, if `evt` was a 1.1 CE then `evt.Xyz = 1;` should fail.

Reply:
The go sdk does this first class support and does it by having whatever is returned from `var evt = new CloudEvent(CloudEventsSpecVersion.V1_0);`, evt is strongly typed as a `v1.0` event. It would have extensions accessors but `evt.Xyz` is not possible. The member did not exist and does not exist in v1.0.

Reply:
Right and that's fine for an official SDK - but we need to remember that anyone who creates code to deal with CEs is like an SDK and whatever rules we define apply to them as well. It also means that anyone who creates custom CE processing code could choose to support `evt.Xyz` as an extension - which means I think it would be wrong for us to mandate that `evt.Xyz = 1` MUST fail.